File: src/models/parsers.ts
import type { HouseholdRecord, SpreadsheetRowInput, VisitOutcome } from "./household";

const COMMENT_NORMALIZATION_RULES: Array<[RegExp, VisitOutcome]> = [
  [/^not\s*home$/i, "not_home"],
  [/^left\s*mess?a?g?e?$/i, "left_message"],
  [/^receptive$/i, "receptive"],
  [/^do\s*not\s*visit$/i, "do_not_visit"],
  [/^moved$/i, "moved"],
  [/^dawat\s*saath$/i, "dawat_saath"],
];

function normalizeText(value: string | undefined): string {
  return (value ?? "").trim();
}

function normalizeOptionalText(value: string | undefined): string | null {
  const normalized = normalizeText(value);
  return normalized.length > 0 ? normalized : null;
}

export function normalizeBriefComment(value: string | undefined): VisitOutcome | string {
  const normalized = normalizeText(value);
  if (!normalized) {
    return "other";
  }

  for (const [pattern, outcome] of COMMENT_NORMALIZATION_RULES) {
    if (pattern.test(normalized)) {
      return outcome;
    }
  }

  return normalized;
}

export function parseLastVisited(value: string | undefined): string | null {
  const normalized = normalizeText(value);
  if (!normalized) {
    return null;
  }

  const fullDateMatch = normalized.match(/^(\d{1,2})\/(\d{1,2})\/(\d{2}|\d{4})$/);
  if (fullDateMatch) {
    const [, monthPart, dayPart, yearPart] = fullDateMatch;
    const month = Number(monthPart);
    const day = Number(dayPart);
    const year = yearPart.length === 2 ? 2000 + Number(yearPart) : Number(yearPart);

    const parsed = new Date(Date.UTC(year, month - 1, day));
    if (
      parsed.getUTCFullYear() === year &&
      parsed.getUTCMonth() + 1 === month &&
      parsed.getUTCDate() === day
    ) {
      return parsed.toISOString().slice(0, 10);
    }
  }

  return null;
}

export function createHouseholdId(input: {
  name: string;
  streetAddress: string;
  neighborhood: string;
}): string {
  const base = `${input.name}|${input.streetAddress}|${input.neighborhood}`
    .toLowerCase()
    .replace(/[^a-z0-9|]+/g, "-")
    .replace(/-+/g, "-")
    .replace(/(^-|-$)/g, "");

  return `household_${base || "unknown"}`;
}

export function parseSpreadsheetRow(
  row: SpreadsheetRowInput,
  context: { sheetName: string; rowNumber: number }
): HouseholdRecord {
  const name = normalizeText(row.name);
  const streetAddress = normalizeText(row.streetAddress);
  const neighborhood = normalizeText(row.neighborhood);

  return {
    id: createHouseholdId({ name, streetAddress, neighborhood }),
    name,
    streetAddress,
    neighborhood,
    briefComment: normalizeBriefComment(row.briefComments),
    lastVisited: parseLastVisited(row.lastVisited),
    notes: normalizeOptionalText(row.notes),
    source: {
      sheetName: context.sheetName,
      rowNumber: context.rowNumber,
    },
    raw: {
      briefComments: row.briefComments,
      lastVisited: row.lastVisited,
      name: row.name,
      streetAddress: row.streetAddress,
      neighborhood: row.neighborhood,
      notes: row.notes,
    },
  };
}

File: android/app/src/main/java/org/outreach/core/model/Parsers.kt
package org.outreach.core.model

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.Locale

private val fullDateRegex = Regex("""^(\d{1,2})\/(\d{1,2})\/(\d{2}|\d{4})$""")
private val normalizationRules = listOf(
    Regex("^not\\s*home$", RegexOption.IGNORE_CASE) to VisitOutcome.NOT_HOME,
    Regex("^left\\s*mess?a?g?e?$", RegexOption.IGNORE_CASE) to VisitOutcome.LEFT_MESSAGE,
    Regex("^receptive$", RegexOption.IGNORE_CASE) to VisitOutcome.RECEPTIVE,
    Regex("^do\\s*not\\s*visit$", RegexOption.IGNORE_CASE) to VisitOutcome.DO_NOT_VISIT,
    Regex("^moved$", RegexOption.IGNORE_CASE) to VisitOutcome.MOVED,
    Regex("^dawat\\s*saath$", RegexOption.IGNORE_CASE) to VisitOutcome.DAWAT_SAATH
)

fun normalizeText(value: String?): String = value?.trim().orEmpty()

fun normalizeOptionalText(value: String?): String? = normalizeText(value).ifBlank { null }

fun normalizeBriefComment(value: String?): String {
    val normalized = normalizeText(value)
    if (normalized.isBlank()) return VisitOutcome.OTHER.name.lowercase(Locale.US)
    val mapped = normalizationRules.firstOrNull { (pattern, _) -> pattern.matches(normalized) }?.second
    return mapped?.name?.lowercase(Locale.US) ?: normalized
}

fun parseLastVisited(value: String?): String? {
    val normalized = normalizeText(value)
    if (normalized.isBlank()) return null
    val match = fullDateRegex.matchEntire(normalized) ?: return null
    val month = match.groupValues[1].toInt()
    val day = match.groupValues[2].toInt()
    val rawYear = match.groupValues[3]
    val year = if (rawYear.length == 2) 2000 + rawYear.toInt() else rawYear.toInt()
    return try {
        LocalDate.of(year, month, day).format(DateTimeFormatter.ISO_DATE)
    } catch (_: DateTimeParseException) {
        null
    }
}

fun createHouseholdId(name: String, streetAddress: String, neighborhood: String): String {
    val base = "$name|$streetAddress|$neighborhood"
        .lowercase(Locale.US)
        .replace(Regex("[^a-z0-9|]+"), "-")
        .replace(Regex("-+"), "-")
        .trim('-')
    return "household_${if (base.isBlank()) "unknown" else base}"
}

fun parseSpreadsheetRow(
    row: SpreadsheetRowInput,
    source: SourceMetadata
): HouseholdRecord {
    val name = normalizeText(row.name)
    val streetAddress = normalizeText(row.streetAddress)
    val neighborhood = normalizeText(row.neighborhood)
    return HouseholdRecord(
        id = createHouseholdId(name, streetAddress, neighborhood),
        name = name,
        streetAddress = streetAddress,
        neighborhood = neighborhood,
        briefComment = normalizeBriefComment(row.briefComments),
        lastVisited = parseLastVisited(row.lastVisited),
        notes = normalizeOptionalText(row.notes),
        source = source,
        raw = RawHouseholdRow(
            briefComments = row.briefComments,
            lastVisited = row.lastVisited,
            name = row.name,
            streetAddress = row.streetAddress,
            neighborhood = row.neighborhood,
            notes = row.notes
        )
    )
}
